src/Database/Esqueleto/PostgreSQL.hs

```diff
@@ -477,3 +486,97 @@ forUpdateOf lockableEntities onLockedBehavior =
 forShareOf :: LockableEntity a => a -> OnLockedBehavior -> SqlQuery ()
 forShareOf lockableEntities onLockedBehavior =
   putLocking $ PostgresLockingClauses [PostgresLockingKind PostgresForShare (Just $ LockingOfClause lockableEntities) onLockedBehavior]
+
+-- | @WITH@ @MATERIALIZED@ clause is used to introduce a
+-- [Common Table Expression (CTE)](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#Common_table_expression)
+-- with the MATERIALIZED keyword. The MATERIALIZED keyword is only supported in PostgreSQL >= version 12.
+-- In Esqueleto, CTEs should be used as a subquery memoization tactic. PostgreSQL treats a materialized CTE as an optimization fence.
+-- A materialized CTE is always fully calculated, and is not "inlined" with other table joins.
+-- Without the MATERIALIZED keyword, PostgreSQL >= 12 may "inline" the CTE as though it was any other join.
+-- You should always verify that using a materialized CTE will in fact improve your performance
+-- over a regular subquery.
+--
+-- @
+-- select $ do
+-- cte <- withMaterialized subQuery
+-- cteResult <- from cte
+-- where_ $ cteResult ...
+-- pure cteResult
+-- @
+--
+--
+-- For more information on materialized CTEs, see the PostgreSQL manual documentation on
+-- [Common Table Expression Materialization](https://www.postgresql.org/docs/14/queries-with.html#id-1.5.6.12.7).
+--
+-- /Since: 3.5.10.0/
+withMaterialized :: ( ToAlias a
```

Comment:
We probably want to also support non materialized so perhaps the materialization modifier can be passed in as an argument and we can provide helpers `materialized` and `notMaterialized` to safely select from. In fact, the more general versions could be in the core module and then whether we export with as `withInternal noModifier` or `withInternal` is whether you use with from the core module or with from this module.

Reply:
So here was what I was thinking with this PR:

- `with`, in Database/Esqueleto/Experimental/From/CommonTableExpression.hs, is
  usable for any of the sql backends; it doesn't do anything with materialized.
  - If a postgres user wants to just use regular `with` CTE and have postgres
    figure out if it wants to materialize it or not, they can just use `with`
    from the normal module.
- the `MATERIALIZED` modifier is only supported in postgres, so the ability to
  specify this behavior is in the pg specific module.
  - If a pg user wishes to force pg to materialize the cte, they must import the
    pg module and use `withMaterialized`
- But, besides `MATERIALIZED`, there is also a `NOT MATERIALIZED` possible
  modifier, which I have *not* added support for in this version . I can add
  it if desired, however, but I'll discuss why I didn't do it below.

Given that, when you say:

> We probably want to also support non materialized

Are you referring to when *no* materialization modifier is specified, or when
the `NOT MATERIALIZED` modifier is intended to be used? Either way is fine; it
makes sense to include something that would do `NOT MATERIALIZED`, for
completeness.

--- 

On the topic of supporting the `NOT MATERIALIZED` modifier, here is what my
thought process was:

I saw the following function pattern when investigating this feature:
- `with` 
- `withRecursive` 

To me, this implied I should add these functions:
- `withMaterialized`
- `withMaterializedRecursive`

But, for `NOT MATERIALIZED`, following the pattern would result in even more
functions (and this just seemed like too many):
- `withNotMaterialized`
- `withNotMaterializedRecursive`

So, I didn't add them; totally fine with adding them, I just want to make sure
we're talking about the same thing. 

--- 

Also, I just re-examined the docs for pg 12-15, and I notice that recursive CTEs are
_always_ "materialized", so `withMaterializedRecursive` and
`withNotMaterializedRecursive` are superfluous. FYI, according to my tests
with pg 15, all of these combinations actually _function_, even if they all
actually behave in the same way. 

---

So, given the above, I *think* the API should be one of these two choices:

- first scenario, I delete `withMaterializedRecursive` and add `withNotMaterialized`
- second scenario, add modifiers as arguments
  - pg module would have
    - `with` function having *two* arguments
      - first argument is `Maybe Materilalization`
        - `data Materilization = Materialized | NotMaterialized`
        - So three possible settings,
          - `None` for providing _no_ materialized modifier
          - `Just Materialized` for providing `MATERIALIZED` modifier
          - `Just NotMaterialized` for providing `NOT MATERIALIZED` modifier
      - the second argument is `query`, as it is today
    - `withRecursive` is deleted 
  - the regular `with` function should have the same API as today in Experimental.From.CommonTableExpression.

Thoughts?

Reply:
@belevy I went ahead and made this change anyway

Reply:
I'm pretty happy with the API that is presented in this PR.